realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -0,0 +1,279 @@
+
+    interface Callback {
+        void onResult(int count);
+    }
+
+    private static class RefAndCount {
+        // The Realm instance in this thread.
+        private final ThreadLocal<BaseRealm> localRealm = new ThreadLocal<BaseRealm>();
+        // How many references to this Realm instance in this thread.
+        private final ThreadLocal<Integer> localCount = new ThreadLocal<Integer>();
+        // How many threads have instances refer to this configuration.
+        private int globalCount = 0;
+    }
+    private enum RealmCacheType {
+        TYPED_REALM,
+        DYNAMIC_REALM;
+
+        static RealmCacheType valueOf(Class<? extends BaseRealm> cls) {
+            if (cls == Realm.class) {
+                return TYPED_REALM;
+            } else if (cls == DynamicRealm.class) {
+                return DYNAMIC_REALM;
+            }
+
+            throw new IllegalArgumentException(WRONG_REALM_CLASS_MESSAGE);
+        }
+    }
+    // Separated references and counters for typed Realm and dynamic Realm.
+    private final EnumMap<RealmCacheType, RefAndCount> refAndCountMap;
+
+    private RealmConfiguration configuration;
+    // Column indices are cached to speed up opening typed Realm. If a Realm instance is created in one thread, creating
+    // Realm instances in other threads doesn't have to initialize the column indices again.
+    private ColumnIndices typedColumnIndices;
+
+    // Realm path will be used as the key to store different RealmCaches. Different Realm configurations with same path
+    // are not allowed and an exception will be thrown when trying to add it to the cache map.
+    private static Map<String, RealmCache> cachesMap = new HashMap<String, RealmCache>();
+
+    private static final String DIFFERENT_KEY_MESSAGE = "Wrong key used to decrypt Realm.";
+    private static final String WRONG_REALM_CLASS_MESSAGE = "The type of Realm class must be Realm or DynamicRealm.";
+
+    private RealmCache(RealmConfiguration config) {
+        configuration = config;
+        refAndCountMap = new EnumMap<RealmCacheType, RefAndCount>(RealmCacheType.class);
+        for (RealmCacheType type : RealmCacheType.values()) {
+            refAndCountMap.put(type, new RefAndCount());
+        }
+    }
+
+    /**
+     * Creates a new Realm instance or get an existing instance for current thread.
+     *
+     * @param configuration {@link RealmConfiguration} will be used to create or get the instance.
+     * @param realmClass class of {@link Realm} or {@link DynamicRealm} to be created in or gotten from the cache.
+     * @return the {@link Realm} or {@link DynamicRealm} instance.
+     */
+    @SuppressWarnings("unchecked")
+    static synchronized <E extends BaseRealm> E createRealmOrGetFromCache(RealmConfiguration configuration,
+                                                        Class<E> realmClass) {
+        boolean isCacheInMap = true;
+        RealmCache cache = cachesMap.get(configuration.getPath());
```

Comment:
we're not checking the thread associated with the `RealmCache`  we got from the map.
different thread should open a new Realm + RealmCache, & cachesMap is not thread local?